DOC-3147: Migration guide from TinyMCE 7 to TinyMCE 8. (#3781)
tinymce/tinymce-docs: modules/ROOT/pages/migration-from-7x.adoc

**Migration checklist:**

* [ ] Identify images previously using `role="presentation"` for decorative purposes in your content
* [ ] Replace those with empty alt attributes (`alt=""`)
* [ ] Update image plugin configuration if customized
* [ ] Test accessibility checker with updated content

==== Page Break plugin update for Export PDF/Word Compatibility
// #TINY-12013

The Page Break plugin has been updated to work out-of-the-box with the xref:exportpdf.adoc[Export to PDF] and xref:exportword.adoc[Export to Word] plugins, addressing a common pain point for developers.

**Impact**: This change improves the export experience by providing more predictable page break behavior.

**Migration steps:**

* If the old behavior is prefered, explicitly set the `pagebreak_separator` option:

[source, javascript]
----
pagebreak_separator: '<!-- my page break -->'
----

The default value has been changed to produce a page break in both the xref:exportpdf.adoc[Export to PDF] and xref:exportword.adoc[Export to Word] service, such as:

.Example
[source, javascript]
----

> pagebreak_separator: '<div class="mce-pagebreak"></div>'

Author: ```suggestion
pagebreak_separator: '<div style="break-after: page"></div>'
```

:description: Guidance for migrating from TinyMCE 7 to TinyMCE 8
:keywords: migration, considerations, premigration, pre-migration
:release-version: 8.0
:page-toclevels: 3

== Introduction

This guide provides a comprehensive overview of the breaking changes introduced in {productname} {release-version}, along with the necessary steps to migrate from {productname} 7.x. It covers key updates to APIs, plugins, and service configurations, including deprecated methods, renamed components, and removed features. These changes are designed to enhance performance, simplify configuration, and align with modern web standards, ensuring a smoother transition and continued compatibility for your integrations.

[IMPORTANT]
.Breaking Changes Quick Reference
====
The following table summarizes all breaking changes in {productname} {release-version}. Each item links to detailed information further in the guide.

[cols="2,3,1",options="header"]
|===

> |Breaking Change |Impact |Severity

Author: Updating to `Breaking Change | Impact | Level`

|`editor.selection.setContent` Deprecated
|Method deprecated. Use `editor.insertContent` instead.
|Medium

|`editor.documentBaseUrl` Removal
|Undocumented property removed. Use `editor.editorManager.documentBaseURL` instead.
|Low

|`skipFocus` and `skip_focus` Consolidation
|Options consolidated to `skipFocus` in `ToggleToolbarDrawer`.
|Low

|`fire()` Method Deprecation
|Method deprecated. Use `dispatch()` for event handling.
|Medium
|===

[TIP]
All items marked "High" severity require immediate attention during migration.
====

== Core API Changes

[discrete]
=== Breaking Changes Overview

> IMPORTANT: The following sections detail critical changes that require updates to your code. Please review each section carefully.

Author: ```suggestion
IMPORTANT: The following sections detail important changes that require updates to your code. Please review each section carefully.
```

**Impact**: Custom language packs using the old naming format will need to be updated.

**Migration checklist:**

* [ ] Identify all language pack files in your deployment
* [ ] Rename files to RFC5646 format (e.g., `en_GB.js` → `en-GB.js`)
* [ ] Update configuration references to language files
* [ ] Update build scripts that handle language files
* [ ] Test language switching in your application
* [ ] Update custom translation files to new format

[source, javascript]
----
// Old configuration
language_url: '/langs/en_GB.js'

// New configuration
language_url: '/langs/en-GB.js'
----

==== Update to Image and Accessibility Checker Plugins
// #TINY-12235

The Image and Accessibility Checker plugins now follow the latest W3C standards for decorative images, requiring an empty alt attribute rather than a `+role="presentation"+` attribute. This change helps improve accessibility support.

**Impact**: Customers using these plugins will need to update their configurations to ensure continued compliance with accessibility standards.

**Migration checklist:**

* [ ] Identify images using `role="presentation"` in your content

> * [ ] Replace with empty alt attributes (`alt=""`)

Author: The language was clarified to specify that only images previously using role="presentation" for decorative purposes need to be updated, not every image.

plugins: 'export',

// Updated configuration plugin identifier
plugins: 'exportlegacy',
----

* Review documentation and user guides to reflect this change.

[TIP]
Alternatively, consider migrating to the new xref:exportpdf.adoc[Export to PDF] and xref:exportword.adoc[Export to Word] plugins for improved functionality and support.

==== Language Pack Filename Changes
// #TINY-12090

Language pack filenames have been standardized to follow the RFC5646 format. This change ensures consistent language handling across different platforms and improves internationalization support.

**Impact**: Custom language packs using the old naming format will need to be updated.

**Migration checklist:**

* [ ] Identify all language pack files in your deployment
* [ ] Rename files to RFC5646 format (e.g., `en_GB.js` → `en-GB.js`)
* [ ] Update configuration references to language files
* [ ] Update build scripts that handle language files
* [ ] Test language switching in your application
* [ ] Update custom translation files to new format

[source, javascript]
----
// Old configuration
language_url: '/langs/en_GB.js'

// New configuration
language_url: '/langs/en-GB.js'

> ----

Author: Yes, you are correct- but should we be advising integrators to transition to use the new format since it will become a breaking changes technically in the new major?

> ----

Reviewer: We should be advising integrators of the need to migrate, but this documentation makes it sound like a breaking change in 8. It's not, it's a deprecation.

> Language pack filenames have been standardized to follow the RFC5646 format. This change ensures consistent language handling across different platforms and improves internationalization support.
> 
> **Impact**: Custom language packs using the old naming format will need to be updated.
> 
> **Migration checklist:**
> 
> * [ ] Identify all language pack files in your deployment
> * [ ] Rename files to RFC5646 format (e.g., `en_GB.js` → `en-GB.js`)
> * [ ] Update configuration references to language files
> * [ ] Update build scripts that handle language files
> * [ ] Test language switching in your application
> * [ ] Update custom translation files to new format
> 
> [source, javascript]
> ----
> // Old configuration
> language_url: '/langs/en_GB.js'
> 
> // New configuration
> language_url: '/langs/en-GB.js'
> ----

Author: ```suggestion
Language pack filenames have been standardized to follow the RFC5646 format. This update ensures consistent language handling across platforms and improves internationalization support. While both the legacy underscore format (e.g., `en_GB.js`) and the new hyphenated format (e.g., `en-GB.js`) are supported in {productname} {release-version}, the underscore format is deprecated and will be removed in {productname} 9. Migrating to the RFC5646 format now will ensure future compatibility and reduce maintenance overhead during upcoming upgrades.

**Impact**: Custom language packs and configurations using the underscore-based format should update before upgrading to {productname} 9 to avoid loading failures.

**Migration checklist:**

* [ ] Identify all language pack files in your deployment
* [ ] Rename files to RFC5646 format (e.g., `en_GB.js` → `en-GB.js`)
* [ ] Update configuration references to language files
* [ ] Update build scripts that handle language files
* [ ] Test language switching in your application
* [ ] Update custom translation files to use the new format

[source, javascript]
----
// Deprecated format (supported only in versions 8.x)
language_url: '/langs/en_GB.js'

// Recommended format
language_url: '/langs/en-GB.js'
----

[IMPORTANT]
Support for the underscore format will be removed in {productname} 9. Early migration is recommended.

```

==== Removed Empty Files
// #TINY-11287, #TINY-12084

Several empty files have been removed from the codebase to reduce clutter and improve maintenance:

* Empty CSS file from the Comments plugin
* Empty LESS file from the Mentions plugin

**Impact**: These changes have no functional impact but may affect custom build processes that explicitly reference these files.

**Migration checklist:**

* [ ] Check build processes for references to Comments plugin CSS
* [ ] Check build processes for references to Mentions plugin LESS
* [ ] Remove any imports of these empty files
* [ ] Test Comments and Mentions plugins after removal

=== Content Handling Updates

==== Improved DIV Element Handling
// #TINY-12171

The handling of `<div>` elements during cut operations has been improved to prevent issues with element structure and content preservation.

**Impact**: This change affects how div elements are handled during cut operations, particularly in complex document structures.

**Migration checklist:**

* [ ] Test cut/paste with nested div structures
* [ ] Update custom div handling code if present
* [ ] Review custom cleanup functions affecting divs

> * [ ] Verify content structure after cut operations

Reviewer: I took another look at the ticket, and I think we can completely remove this one from the migration guide.

It's a fix to make cut behaviour match user expectations. Previously "cut" on a `div` element would not completely remove it, except in table cells. Now it is always removed.

> * [ ] Verify content structure after cut operations

Reviewer: correct, delete it all

* **`SAFE_FOR_XML` enabled** — This setting enforces stricter handling of comments and attribute values, preventing certain XSS vectors.
* **Content Impact** — HTML comments containing tags, Internet Explorer conditional comments, and attributes with HTML-like values may now be removed during sanitization. Content that was previously allowed may be stripped.

==== Migration Guidance

* Review workflows and test content that previously relied on relaxed sanitization.
* If custom comment handling is required, consider using DOMPurify hooks as recommended by the maintainers.

.Example: Content Differences
|===
|Content |{productname} 7 Output |{productname} {release-version} Output

|`<div><!-- <p>Hello World</p> --></div>`
|`<div><!-- <p>Hello World</p> --></div>`
|`<div></div>`

|`<iframe><!-- Hello World --></iframe>`
|`<p><iframe sandbox=""><!-- Hello World --></iframe></p>`
|``

|`<script><!-- Hello World --></script>`
|`<script><!-- Hello World --></script>`
|``
|===

[NOTE]
====
For information on disabling DOMPurify sanitization (not recommended), see xref:security.adoc#xss_sanitization-option[xss_sanitization option].
====

> 

Reviewer 2: ```suggestion
== Core Changes

  === Split Button CSS Structure Updates

  If you have custom CSS targeting split button
  internals, you may need to update your selectors to
  accommodate the new structure.

  Split buttons now render as two separate components
  with updated CSS classes:

  * `.tox-split-button__main` - Main action button
  * `.tox-split-button__chevron` - Dropdown chevron
  button

  ==== Migration Required

  Custom CSS that targets split button internals may
  need to be updated to reflect these changes.

  [source,css]
  ----
  /* Before TinyMCE 8.0 */
  .tox-split-button {
    /* Previous styling approach */
  }

  /* TinyMCE 8.0+ */
  .tox-split-button__main {
    /* Main button styling */
  }

  .tox-split-button__chevron {
    /* Chevron button styling */
  }
  ----
```

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3781/commits/a878bf8cc8d9aaf675f2bdb9dbeccbf80ab656b4) @ArvinJ-H see updated changes.

Reviewer: I disagree.

=== Deprecated Methods

[[fire-method-deprecation]]
==== `fire()`
// #TINY-11692

The `fire()` method has been replaced by `dispatch()` for event handling. The `fire()` method will be removed in {productname} 9 to avoid confusion with its name.

[source, javascript]
----
// Deprecated in TinyMCE 8, will be removed in 9
// Old approach for dispatching custom events
editor.fire('someEvent');

// New approach for dispatching custom events
editor.dispatch('someEvent');
----

**Impact**: This change aligns {productname} with modern event handling conventions, making the API more intuitive for developers.

**Migration checklist:**

* [ ] Search codebase for all uses of the `fire()` method
* [ ] Replace each instance with `dispatch()`
* [ ] Review and update third-party plugins
* [ ] Test all custom event handling

=== Plugin Updates

==== List Plugin Updates
// #TINY-11937

The list plugin now includes configurable indentation level restrictions to prevent excessive nesting of list items. This change helps maintain cleaner document structure and improves compatibility with various export formats.

[source, javascript]
----
// Configure maximum list indentation levels
tinymce.init({

>   max_list_indent: 3  // Restricts list indentation to 3 levels

Author: ```suggestion
  list_max_length: 3  // Restricts list indentation to 3 levels
```

});
----

**Impact**: Lists will no longer indent beyond the configured maximum level.

**Migration checklist:**

* [ ] Audit existing content for deeply nested lists
* [ ] Determine appropriate maximum indentation level for your use case
* [ ] Configure `max_list_indent` setting

> * [ ] Test list behavior with existing content

Author: noted, thanks for picking up on this.

> * [ ] Test list behavior with existing content

Author: since this is not breaking change, ill remove it - this [ticket](https://github.com/tinymce/tinymce-docs/pull/3766/files) covers the new option and release note entry.

> ==== List Plugin Updates
> // #TINY-11937
> 
> The list plugin now includes configurable indentation level restrictions to prevent excessive nesting of list items. This change helps maintain cleaner document structure and improves compatibility with various export formats.
> 
> [source, javascript]
> ----
> // Configure maximum list indentation levels
> tinymce.init({
>   list_max_length: 3  // Restricts list indentation to 3 levels
> });
> ----
> 
> **Impact**: Lists will no longer indent beyond the configured maximum level.
> 
> **Migration checklist:**
> 
> * [ ] Audit existing content for deeply nested lists
> * [ ] Determine appropriate maximum indentation level for your use case
> * [ ] Configure `max_list_indent` setting
> * [ ] Test list behavior with existing content
> 

Author: ```suggestion
```

**Impact**: This change affects how page breaks interact with block elements, providing a more intuitive editing experience.

**Migration steps:**

* If you want to maintain the previous behavior where page breaks do not automatically split block elements, add the following to your configuration:

[source, javascript]
----
tinymce.init({
  // ...other configuration options...
  pagebreak_split_block: false
});
----

=== Technical Improvements and Cleanup

==== Cross-Origin Resource Loading Configuration
// #TINY-12228, TINY-12326

When upgrading to {productname} 8, you will need to review and possibly update how your integration handles cross-origin resource loading. {productname} 8 provides a new configuration option for controlling the `crossorigin` attribute on loaded resources.

**What to check:**

. If you're using {cloudname}:
+
* Ensure your script tag includes both required attributes:
+
[source, html]
----

> <script src="{cdnurl}" referrerpolicy="origin" crossorigin="anonymous"></script>

Author: Was missing `[source,html,subs="attributes+"]`

> <script src="{cdnurl}" referrerpolicy="origin" crossorigin="anonymous"></script>

Author: fixed in [this commit](https://github.com/tinymce/tinymce-docs/pull/3781/commits/d1bc75f5322add15b4b86c4555b3df600ecb84d2)

**License Migration checklist:**

* [ ] Contact support for new {productname} {release-version} license key or use GPL for the open source version
* [ ] Install license key manager addon for commercial licenses
* [ ] Update configuration with new license key format
* [ ] Test editor functionality with new license
* [ ] Verify all premium features are working

[[dompurify-update-breaking-change]]
=== DOMPurify Update and Stricter Sanitization (Breaking Change)
// #TINY-12056

{productname} {release-version} updates the DOMPurify dependency to version 3.2.6 and enables the `SAFE_FOR_XML` flag by default. This is a breaking change: content that previously passed sanitization in {productname} 7 may now be stripped or altered during the sanitization process.

[IMPORTANT]
====
This change improves security and aligns with DOMPurify's recommended defaults. However, existing content and integrations that relied on the previous, less strict sanitization behavior may be impacted.
====

==== Key Changes

> * **DOMPurify upgraded to 3.2.6** — The custom patch is now included upstream; maintaining a forked patch is no longer required.

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3781/commits/a878bf8cc8d9aaf675f2bdb9dbeccbf80ab656b4)

* **`SAFE_FOR_XML` enabled** — This setting enforces stricter handling of comments and attribute values, preventing certain XSS vectors.
* **Content Impact** — HTML comments containing tags, Internet Explorer conditional comments, and attributes with HTML-like values may now be removed during sanitization. Content that was previously allowed may be stripped.

==== Migration Guidance

* Review workflows and test content that previously relied on relaxed sanitization.

> * If custom comment handling is required, consider using DOMPurify hooks as recommended by the maintainers.

Author: But ill update this section to include the feedback also, thanks

> * If custom comment handling is required, consider using DOMPurify hooks as recommended by the maintainers.

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3781/commits/a878bf8cc8d9aaf675f2bdb9dbeccbf80ab656b4)

* Review workflows and test content that previously relied on relaxed sanitization.
* {productname} now provides the xref:content-filtering.adoc#allow_html_in_comments[Content Filtering: allow_html_in_comments option] option. Enabling this option allows HTML tags in comments with sanitization still enabled. 

[WARNING]
Using `allow_html_in_comments` increases the risk of XSS vulnerabilities. xref:security.adoc#allow_html_in_comments[allow_html_in_comments] is not recommended for production use unless you fully understand the implications and have appropriate security measures in place.

.Example: Content Differences
|===
|Content |{productname} 7 Output |{productname} {release-version} Output

|`<div><!-- <p>Hello World</p> --></div>`
|`<div><!-- <p>Hello World</p> --></div>`
|`<div></div>`

|`<iframe><!-- Hello World --></iframe>`
|`<p><iframe sandbox=""><!-- Hello World --></iframe></p>`
|``

|`<script><!-- Hello World --></script>`
|`<script><!-- Hello World --></script>`
|``
|===

[NOTE]
====
For information on disabling DOMPurify sanitization (not recommended), see xref:security.adoc#xss_sanitization-option[xss_sanitization option].
====

== Core Changes

> [[split-button-css-breaking-change]]
> === Split button CSS structure updates
> // #TINY-8665
> 
> {productname} {productmajorversion} updates the internal structure of split buttons. Split buttons now render as two separate components for the main action and the dropdown chevron.
> 
> **Impact**: This change only affects users with custom skins.
> 
> **Migration steps**:
> 
> If you have a custom skin, rebuild it using the {productname} {productmajorversion} codebase. See xref:creating-a-skin.adoc[Creating a Skin] for guidance on the custom skin development process.
> 
> **Migration checklist:**
> 
> * [ ] Identify if you are using a custom skin.
> * [ ] Rebuild your custom skin using the {productname} {productmajorversion} codebase.
> * [ ] Test your custom skin with split buttons in {productname} {productmajorversion}.
> * [ ] Refer to xref:creating-a-skin.adoc[Creating a Skin] for details.

Author: ```suggestion
[[split-button-css-breaking-change]]
=== Split button CSS structure updates
// #TINY-8665

{productname} {productmajorversion} updates the internal structure of split buttons. Split buttons now render as two separate components for the main action and the dropdown chevron.

**Impact**: This change only affects users with custom skins.

**Migration steps**:

If you have a custom skin, rebuild it using the {productname} {productmajorversion} codebase. See xref:creating-a-skin.adoc[Creating a Skin] for guidance on the custom skin development process.

**Migration checklist:**

* [ ] Identify if you are using a custom skin.
* [ ] Rebuild your custom skin using the {productname} {productmajorversion} codebase.
* [ ] Test your custom skin with split buttons in {productname} {productmajorversion}.
* [ ] Refer to xref:creating-a-skin.adoc[Creating a Skin] for details.

**Code examples**

.Before: {productname} {productmajorversion}.
[source,js]
----
tinymce.init({
  selector: 'textarea#custom-toolbar-split-button',
  toolbar: 'myButton',
  menubar: false,
  setup: (editor) => {
    editor.ui.registry.addSplitButton('myButton', {
      text: 'My Button',
      icon: 'info',
      tooltip: 'This is an example split-button',
      onAction: () => editor.insertContent('<p>You clicked the main button</p>'),
      onItemAction: (api, value) => editor.insertContent(value),
      fetch: (callback) => {
        const items = [
          {
            type: 'choiceitem',
            icon: 'notice',
            text: 'Menu item 1',
            value: '&nbsp;<em>You clicked menu item 1!</em>'
          },
          {
            type: 'choiceitem',
            icon: 'warning',
            text: 'Menu item 2',
            value: '&nbsp;<em>You clicked menu item 2!</em>'
          }
        ];
        callback(items);
      }
    });
  },
  content_style: 'body { font-family:Helvetica,Arial,sans-serif; font-size:16px }'
});
----

.Migrating to {productname} {productmajorversion}
[source,js]
----
tinymce.init({
  selector: 'textarea',
  toolbar: 'myButton',
  menubar: false,
  setup: (editor) => {
    editor.ui.registry.addSplitButton('myButton', {
      text: 'My Button',
      icon: 'info',
      tooltip: 'Execute my action',
      chevronTooltip: 'My Button menu options',
      onAction: () => editor.insertContent('<p>You clicked the main button</p>'),
      onItemAction: (api, value) => editor.insertContent(value),
      fetch: (callback) => {
        const items = [
          {
            type: 'choiceitem',
            icon: 'notice',
            text: 'Menu item 1',
            value: '&nbsp;<em>You clicked menu item 1!</em>'
          },
          {
            type: 'choiceitem',
            icon: 'warning',
            text: 'Menu item 2',
            value: '&nbsp;<em>You clicked menu item 2!</em>'
          }
        ];
        callback(items);
      }
    });
  },
  content_style: 'body { font-family:Helvetica,Arial,sans-serif; font-size:16px }'
});
----

```

[IMPORTANT]
.Breaking Changes Quick Reference
====
The following table summarizes all breaking changes in {productname} {release-version}. Each item links to detailed information further in the guide.

Any items marked **"High"** level require immediate attention during migration.

[cols="2,3,1",options="header"]
|===
|Breaking Change |Impact |Level
|xref:license-key-system-update[License Key System]
|Self-hosted deployments now require a new license key format and license key manager. Old keys are **not compatible**.
|High

|xref:dompurify-update-breaking-change[DOMPurify Update]
|Sanitization is now stricter; content previously allowed may be stripped or altered.
|High

|xref:editor-selection-setcontent-deprecated[editor.selection.setContent]
|Method deprecated. Use `editor.insertContent` instead.
|Medium

|xref:fire-method-deprecation[fire()]
|Method deprecated. Use `dispatch()` for event handling.
|Medium

> <<<<<<< Updated upstream
> |xref:pagebreak-split-block-default[pagebreak_split_block]
> |The default value changed from `false` to `true`, affecting how page breaks interact with block elements.
> |Medium
> 
> =======
> >>>>>>> Stashed changes

Author: ```suggestion
```

|``

|`<script><!-- Hello World --></script>`
|`<script><!-- Hello World --></script>`
|``
|===

[NOTE]
====
For information on disabling DOMPurify sanitization (not recommended), see xref:security.adoc#xss_sanitization-option[xss_sanitization option].
====

== Core Changes

[[split-button-css-breaking-change]]
=== Split button CSS structure updates
// #TINY-8665

{productname} {productmajorversion} updates the internal structure of split buttons. Split buttons now render as two separate components for the main action and the dropdown chevron.

**Impact**: This change only affects users with custom skins.

**Migration steps**:

If you have a custom skin, rebuild it using the {productname} {productmajorversion} codebase. See xref:creating-a-skin.adoc[Creating a Skin] for guidance on the custom skin development process.

**Migration checklist:**

* [ ] Identify if you are using a custom skin.
* [ ] Rebuild your custom skin using the {productname} {productmajorversion} codebase.
* [ ] Test your custom skin with split buttons in {productname} {productmajorversion}.
* [ ] Refer to xref:creating-a-skin.adoc[Creating a Skin] for details.

> 
> **Code examples**

Author: ```suggestion
```

[IMPORTANT]
.Breaking Changes Quick Reference
====
The following table summarizes all breaking changes in {productname} {release-version}. Each item links to detailed information further in the guide.

Any items marked **"High"** level require immediate attention during migration.

[cols="2,3,1",options="header"]
|===
|Breaking Change |Impact |Level
|xref:license-key-system-update[License Key System]
|Self-hosted deployments now require a new license key format and license key manager. Old keys are **not compatible**.
|High

|xref:dompurify-update-breaking-change[DOMPurify Update]
|Sanitization is now stricter; content previously allowed may be stripped or altered.
|High

|xref:editor-selection-setcontent-deprecated[editor.selection.setContent]
|Method deprecated. Use `editor.insertContent` instead.
|Medium

|xref:fire-method-deprecation[fire()]
|Method deprecated. Use `dispatch()` for event handling.
|Medium

|xref:pagebreak-split-block-default[pagebreak_split_block]
|The default value changed from `false` to `true`, affecting how page breaks interact with block elements.

> |Low

Reviewer: I think this should be `medium`, and also see my comments about page break below

[IMPORTANT]
====
This change improves security and aligns with DOMPurify's recommended defaults. However, existing content and integrations that relied on the previous, less strict sanitization behavior may be impacted.
====

==== Key Changes

* **DOMPurify upgraded to 3.2.6**
* **`SAFE_FOR_XML` enabled** — This setting enforces stricter handling of comments and attribute values, preventing certain XSS vectors.
* **Content Impact** — HTML comments containing tags, Internet Explorer conditional comments, and attributes with HTML-like values may now be removed during sanitization. Content that was previously allowed may be stripped.

*Impact*: This change improves security by preventing potential XSS attacks through comments and attributes that were previously allowed. However, it may also result in content being stripped or altered unexpectedly.

**Migration Steps:**

* Review workflows and test content that previously relied on relaxed sanitization.

> * If custom comment handling is required, note that DOMPurify hooks cannot be used directly unless you fork TinyMCE or disable TinyMCE's built-in sanitization and implement your own with DOMPurify.

Reviewer 3: Sorry I wasn't  clear previously. I meant that it's not reasonable to mention DOMPurify's hooks at all.
```suggestion
```

== Core Changes

[[split-button-css-breaking-change]]
=== Split button CSS structure updates
// #TINY-8665

{productname} {release-version} updates the internal structure of split buttons. Split buttons now render as two separate components with updated CSS classes:

* `.tox-split-button__main` — Main action button
* `.tox-split-button__chevron` — Dropdown chevron button

**Impact**: This change improves the flexibility of split button styling and allows for more granular control over each component.

**Migration steps:**

Update all CSS selectors and styles that previously targeted the unified `.tox-split-button` class to target the new `.tox-split-button__main` and `.tox-split-button__chevron` classes.

[source,css]
----
/* Before TinyMCE 8.0 */
.tox-split-button {
  /* Previous styling approach */
}

/* TinyMCE 8.0+ */
.tox-split-button__main {
  /* Main button styling */
}

.tox-split-button__chevron {
  /* Chevron button styling */
}
----

**Migration checklist:**

* [ ] Search your CSS for `.tox-split-button` and update to use `.tox-split-button__main` and `.tox-split-button__chevron`.
* [ ] Test all split button styles in your application to ensure they render correctly.

> * [ ] Update any custom plugins that use split buttons to reflect the new structure.

Reviewer: This is all wrong. We actively discourage customers from applying `tox` styles to their page. That may change in the future, but for now we need to stick to our recommendations.

The only customers who will be impacted are those with custom skins. And the only migration step that I expect they need is to rebuild their custom skins using the TinyMCE 8 codebase (which is true of most releases and should be a process they are accustomed to). The custom skin process should be in the docs somewhere, we can link to that.

The chance that someone is using a custom skin _and_ configured the split button specifically is so small it's not worth calling out.

Reviewer 2: ah yes, sorry I got it mixed up again, how about this

```
=== Split button CSS structure updates

  TinyMCE 8.0 updates the internal structure of split
  buttons.

  **Impact**: This change only affects users with
  custom skins.

  **Migration step**: If you have a custom skin,
  rebuild it using the TinyMCE 8.0 codebase. See
  xref:creating-a-skin.adoc[Creating a Skin] for
  guidance on the custom skin development process.
```

Reviewer: That's a better general direction, but I think the step can expand to a checklist

Author: @ArvinJ-H I have updated this section based on your suggestion and @TheSpyder comments, please review.

// #TINY-12226

The Image and Accessibility Checker plugins now follow the latest W3C standards for decorative images, requiring an empty alt attribute rather than a `+role="presentation"+` attribute. This change helps improve accessibility support.

**Impact**: Customers using these plugins will need to update their configurations to ensure continued compliance with accessibility standards.

**Migration checklist:**

* [ ] Identify images previously using `role="presentation"` for decorative purposes in your content
* [ ] Replace those with empty alt attributes `alt=""`
* [ ] Update image plugin configuration if customized
* [ ] Test accessibility checker with updated content

For more information on the changes, see: xref:a11ychecker.adoc##image-rules[Accessibility Checker: Image rules].

[[pagebreak-split-block-default]]
==== Page Break Plugin: Export PDF/Word Compatibility and Option Updates
// #TINY-12013, #TINY-12462

The Page Break plugin has been updated to work out-of-the-box with the xref:exportpdf.adoc[Export to PDF] and xref:exportword.adoc[Export to Word] plugins, addressing a common pain point for developers. As part of this update:

> * The default value of the xref:pagebreak.adoc#pagebreak_separator[`pagebreak_separator`] option now uses a block tag (`<div style="break-after: page

Reviewer: The separator change should be the focus of the breaking change, not split block. Add a new `pagebreak_separator` entry to the table at the top and it is _definitely_ a `medium`. It's almost a `high` but it won't break content on load, so maybe put separator as the first medium and split block as the second.

I suggest linking both entries to this documentation section 🤔

|Low

|xref:split-button-css-breaking-change[Split buttons]
|Split button CSS classes and structure have changed.
|Low
|===
====

[[license-key-system-update]]
=== Transition from Version 7 License Keys to Version {release-version} License Keys
// #EPIC-192

[IMPORTANT]
====
This section applies to self-hosted installations only. For cloud deployments, license key management is handled automatically.
====

{productname} {release-version} introduces an enhanced license key system that requires specific attention during migration.

The complete licensing documentation xref:license-key.adoc[License Key Management] covers:

* Detailed explanations of all license types (GPL, Commercial, GPL with Premium Features)
* Time-based vs Version-locked license key differences
* License states (Active, Grace Period, Expired, Invalid)
* Deployment options (Cloud-only, Self-hosted, Hybrid)
* Step-by-step configuration examples for each setup
* Commercial License Key Manager setup and requirements
* Troubleshooting and FAQ

**Impact**: The new license key system introduces breaking changes that require updates to your configuration and code.

**Key Migration Considerations**:

> * *License Key Format Change:* Version 7 keys are:

Reviewer 4: Should we move “Version 7 keys are:” into the list item? It might make it clearer that the next points belong under `License Key Format Change`.

|``

|`<script><!-- Hello World --></script>`
|`<script><!-- Hello World --></script>`
|``
|===

[NOTE]
====
For information on disabling DOMPurify sanitization (not recommended), see xref:security.adoc#xss_sanitization-option[xss_sanitization option].
====

== Core Changes

[[split-button-css-breaking-change]]
=== Split button CSS structure updates
// #TINY-8665

{productname} {productmajorversion} updates the internal structure of split buttons. Split buttons now render as two separate components for the main action and the dropdown chevron.

**Impact**: This change only affects users with custom skins.

**Migration steps**:

If you have a custom skin, rebuild it using the {productname} {productmajorversion} codebase. See xref:creating-a-skin.adoc[Creating a Skin] for guidance on the custom skin development process.

**Migration checklist:**

* [ ] Identify if you are using a custom skin.
* [ ] Rebuild your custom skin using the {productname} {productmajorversion} codebase.
* [ ] Test your custom skin with split buttons in {productname} {productmajorversion}.
* [ ] Refer to xref:creating-a-skin.adoc[Creating a Skin] for details.

> **Code examples**

Reviewer 4: I think the examples may not be relevant to the migration steps for split buttons (adding a `chevronTooltip` from what I can see). Should we add a heading here to indicate that this should be the new structure for split buttons? Or should we move this to the split button page?

Author: removed as discussed.

.Before: {productname} {productmajorversion}.
[source,js]
----
tinymce.init({
  selector: 'textarea#custom-toolbar-split-button',
  toolbar: 'myButton',
  menubar: false,
  setup: (editor) => {
    editor.ui.registry.addSplitButton('myButton', {
      text: 'My Button',
      icon: 'info',
      tooltip: 'This is an example split-button',
      onAction: () => editor.insertContent('<p>You clicked the main button</p>'),
      onItemAction: (api, value) => editor.insertContent(value),
      fetch: (callback) => {

>         const items = [
>           {
>             type: 'choiceitem',
>             icon: 'notice',
>             text: 'Menu item 1',
>             value: '&nbsp;<em>You clicked menu item 1!</em>'
>           },
>           {
>             type: 'choiceitem',
>             icon: 'warning',
>             text: 'Menu item 2',
>             value: '&nbsp;<em>You clicked menu item 2!</em>'
>           }
>         ];
>         callback(items);
>       }
>     });
>   },
>   content_style: 'body { font-family:Helvetica,Arial,sans-serif; font-size:16px }'
> });
> ----
> 
> .Migrating to {productname} {productmajorversion}
> [source,js]
> ----
> tinymce.init({
>   selector: 'textarea',
>   toolbar: 'myButton',
>   menubar: false,
>   setup: (editor) => {
>     editor.ui.registry.addSplitButton('myButton', {
>       text: 'My Button',
>       icon: 'info',
>       tooltip: 'Execute my action',
>       chevronTooltip: 'My Button menu options',
>       onAction: () => editor.insertContent('<p>You clicked the main button</p>'),
>       onItemAction: (api, value) => editor.insertContent(value),
>       fetch: (callback) => {
>         const items = [
>           {
>             type: 'choiceitem',
>             icon: 'notice',
>             text: 'Menu item 1',
>             value: '&nbsp;<em>You clicked menu item 1!</em>'
>           },
>           {
>             type: 'choiceitem',
>             icon: 'warning',
>             text: 'Menu item 2',
>             value: '&nbsp;<em>You clicked menu item 2!</em>'
>           }
>         ];
>         callback(items);
>       }
>     });
>   },
>   content_style: 'body { font-family:Helvetica,Arial,sans-serif; font-size:16px }'
> });
> ----
> 

Author: ```suggestion
```

== Core API Changes

[discrete]
=== Breaking Changes Overview

IMPORTANT: The following sections detail important changes that require updates to your code. Please review each section carefully.

=== Updated Methods

[[skipfocus-consolidation]]
==== skipFocus and skip_focus
// #TINY-12044

The `skipFocus` and `skip_focus` options for the `ToggleToolbarDrawer` command have been consolidated into a single, more consistent argument. This reduces API complexity and clarifies the intended behavior.

**Impact**: This change simplifies focus management, reducing the risk of confusion and unexpected behavior.

**Migration steps:**

[source, javascript]
----
// Old approach (Deprecated) in TinyMCE 8

> editor.execCommand('ToggleToolbarDrawer', { skip_focus: false }, { skipFocus: true });

Reviewer 4: ```suggestion
editor.execCommand('ToggleToolbarDrawer', false, { skipFocus: true })
```
The command doesn't look correct 🤔  
https://ephocks.atlassian.net/browse/TINY-12044
https://fiddle.tiny.cloud/ClRQ2cMIQk/1

// New approach (Recommended)

> editor.execCommand('ToggleToolbarDrawer', false, { skipFocus: true });

Reviewer 4: ```suggestion
editor.execCommand('ToggleToolbarDrawer', false, null, { skip_focus: true })
```

----

**Migration checklist:**

* [ ] Locate all instances of `ToggleToolbarDrawer` command usage
* [ ] Replace `skip_focus` with `skipFocus` in command options
* [ ] Update any custom plugins using this command
* [ ] Test toolbar drawer behavior after changes

'''

=== Removed Methods

[[editor-documentbaseurl-removal]]
==== editor.documentBaseUrl
[.discrete]
// #TINY-12182

The undocumented `editor.documentBaseUrl` property has been removed.

.Example Usage
[source,javascript]
----
// Removed in TinyMCE 8
console.log('documentBaseUrl', editor.documentBaseUrl);

// Use this instead
console.log('documentBaseURI', editor.documentBaseURI.getURI());
----

> TIP: Use `tinymce.activeEditor.documentBaseURI.getURI()` for all base URL operations.

Author: ```suggestion
TIP: Use `editor.documentBaseURI.getURI()` for all base URL operations.
```

**Impact**: This change improves URL handling consistency by removing an undocumented API that was not aligned with the documented `documentBaseURI` property.

**Migration steps:**

> To update all references of `documentBaseUrl` to the new API, replace any usage of `editor.documentBaseUrl` (or similar) with `tinymce.activeEditor.documentBaseURI.getURI()`. The property `documentBaseUrl` has been removed, and the correct way to access the document base URL is now through the `documentBaseURI` property, which is a URI object. You can then call `.getURI()` on it to get the string value of the URL.

Reviewer 4: ```suggestion
To update all references of `documentBaseUrl` to the new API, replace any usage of `editor.documentBaseUrl` (or similar) with `editor.documentBaseURI.getURI()`. The property `documentBaseUrl` has been removed, and the correct way to access the document base URL is now through the `documentBaseURI` property, which is a URI object. You can then call `.getURI()` on it to get the string value of the URL.
```
Maybe for consistency
